### PR TITLE
A bunch of cool features from the dev branch

### PR DESCRIPTION
**Pull request by lajp**

### CODE REVIEW

File: src/database.rs [hunk 1]
```diff
@@ -3,761 +3,788 @@ use argon2::{
...
-        .select(id)
-        .first::<i32>(conn)
-        .optional()?
-        .is_some())
-}
+    fn delete_activity(&mut self, userid: i32, activity: i32) -> Result<bool, TimeError> {
+        use crate::schema::coding_activities::dsl::*;
+        let res = diesel::delete(crate::schema::coding_activities::table)
+            .filter(id.eq(activity))
+            .filter(user_id.eq(userid))
```
Comment (lajp) [inline src/database.rs:440]: Aren't activity ids unique? If yes, then why are we using `user_id` here?

**Review by lajp — COMMENTED** [review]
@DrVilepis please answer my comment. otherwise this should be able to be merged

File: src/database.rs (hunk 1 above)
Comment (DrVilepis) [reply src/database.rs:440]: They are unique, but this ensure a user cant delete other user's activities

**Review by DrVilepis — APPROVED** [review]
Yep cock

**Comment by DrVilepis** [comment]
This seems good to go so merging now

**Pull request merged**